client/service/src/builder.rs

```diff
@@ -691,55 +675,43 @@ async fn transaction_notifications<TBl, TExPool>(
 		.await;
 }
 
-fn build_telemetry<TBl: BlockT>(
+fn init_telemetry<TBl: BlockT, TCl: BlockBackend<TBl>>(
 	config: &mut Configuration,
-	endpoints: sc_telemetry::TelemetryEndpoints,
-	telemetry_connection_sinks: TelemetryConnectionSinks,
 	network: Arc<NetworkService<TBl, <TBl as BlockT>::Hash>>,
-	spawn_handle: SpawnTaskHandle,
-	genesis_hash: <TBl as BlockT>::Hash,
-) -> sc_telemetry::Telemetry {
+	client: Arc<TCl>,
+) -> Option<sc_telemetry::TelemetryConnectionNotifier> {
+	let telemetry_handle = config.telemetry_handle.take()?;
+
+	let endpoints = match config.telemetry_endpoints.clone() {
+		// Don't initialise telemetry if `telemetry_endpoints` == Some([])
+		Some(endpoints) if !endpoints.is_empty() => endpoints,
+		_ => return None,
+	};
+
+	let genesis_hash = client.block_hash(Zero::zero()).ok().flatten().unwrap_or_default();
+
 	let is_authority = config.role.is_authority();
 	let network_id = network.local_peer_id().to_base58();
 	let name = config.network.node_name.clone();
 	let impl_name = config.impl_name.clone();
 	let impl_version = config.impl_version.clone();
 	let chain_name = config.chain_spec.name().to_owned();
-	let telemetry = sc_telemetry::init_telemetry(sc_telemetry::TelemetryConfig {
-		endpoints,
-		wasm_external_transport: config.telemetry_external_transport.take(),
-	});
 	let startup_time = SystemTime::UNIX_EPOCH.elapsed()
 		.map(|dur| dur.as_millis())
 		.unwrap_or(0);
 
-	spawn_handle.spawn(
-		"telemetry-worker",
-		telemetry.clone()
-			.for_each(move |event| {
-				// Safe-guard in case we add more events in the future.
-				let sc_telemetry::TelemetryEvent::Connected = event;
-
-				telemetry!(SUBSTRATE_INFO; "system.connected";
-					"name" => name.clone(),
-					"implementation" => impl_name.clone(),
-					"version" => impl_version.clone(),
-					"config" => "",
-					"chain" => chain_name.clone(),
-					"genesis_hash" => ?genesis_hash,
-					"authority" => is_authority,
-					"startup_time" => startup_time,
-					"network_id" => network_id.clone()
-				);
-
-				telemetry_connection_sinks.0.lock().retain(|sink| {
-					sink.unbounded_send(()).is_ok()
-				});
-				ready(())
-			})
-	);
-
-	telemetry
+	let mut obj = serde_json::Map::new();
```

Comment:
You had a rust struct before, why did you changed this?...

Reply:
I don't remember there was ever a struct here :thinking: though it is a good idea, I will make a new struct for these parameters! Thanks!

Reply:
Done in b33628b09